### PR TITLE
HPCC-28115 Expose k8s connections for predefined k8s services

### PR DESCRIPTION
**Pull request by wangkx**
Expose service type, IPs and ports for a k8s service if the
annotation hpcc.eclwatch.io/enabled for the service is configured
to be true.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@GordonSmith please review, especially on the interface defined in the ecm file.

File: esp/scm/ws_resources.ecm [hunk 1]
```diff
@@ -25,11 +25,35 @@ ESPStruct HPCCService
...
 };
 
+ESPStruct ServicePorts
+{
+    string Name;
+    string Protocol;
+    unsigned Port;
+    unsigned NodePort;
+    unsigned TargetPort;
+};
```
Comment (GordonSmith) [inline esp/scm/ws_resources.ecm:35]: What are the 3 ports?  I don't think there is any need to know any of the "internal" ports info?

Comment (afishbeck) [reply esp/scm/ws_resources.ecm:35]: I think you can remove the internal TargetPort at least until someone asks for it.  Whether anyone will use the NodePort if the type is loadbalancer I'm not sure.

File: esp/scm/ws_resources.ecm [hunk 2]
```diff
@@ -25,11 +25,35 @@ ESPStruct HPCCService
...
+//The value is based on the existing https://github.com/kubernetes/kubernetes.
+ESPStruct ServiceConnection
+{
+    string Type;
+    ESParray<string> ExternalIPs;
+    string ExternalIPsEx; //Use this field when ExternalIPs is empty. Set to '<none>' for spec type 'ClusterIP' and spec type 'NodePort';
+                          //set to '<pending>' for spec type 'LoadBalancer'; leave it empty for spec type 'ExternalName';
+                          //set to '<unknown>' for unknown spec types.
+    ESParray<string> ClusterIPs;
+    string ClusterIPsEx; //when ClusterIPs is empty, set to '<none>'.
```
Comment (afishbeck) [inline esp/scm/ws_resources.ecm:47]: The Ex on the name is a little confusing.  I tend to think of Ex as an extended version of something.
If the only purpose is to say "<none>" when the IP list is empty, is it really needed?

File: esp/scm/ws_resources.ecm [hunk 3]
```diff
@@ -25,11 +25,35 @@ ESPStruct HPCCService
...
+    unsigned TargetPort;
+};
+
+//The ExternalIPsEx or ClusterIPsEx is used when ExternalIPs or ClusterIPs is empty.
+//The value is based on the existing https://github.com/kubernetes/kubernetes.
+ESPStruct ServiceConnection
+{
+    string Type;
+    ESParray<string> ExternalIPs;
+    string ExternalIPsEx; //Use this field when ExternalIPs is empty. Set to '<none>' for spec type 'ClusterIP' and spec type 'NodePort';
```
Comment (afishbeck) [inline esp/scm/ws_resources.ecm:43]: Also not sure an Ex ending is right here.  Is this field really helpful if it just tells us that the ips are empty?

File: esp/scm/ws_resources.ecm (hunk 2 above)
Comment (wangkx) [reply esp/scm/ws_resources.ecm:47]: I may remove it. We may add a field if the value could not be '<none>'.

File: esp/scm/ws_resources.ecm (hunk 1 above)
Comment (wangkx) [reply esp/scm/ws_resources.ecm:35]: I could be wrong, but I think the NodePort should be used if the type is "NodePort" (not "LoadBalancer"). See:https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/apis/core/types.go#L3475

File: esp/scm/ws_resources.ecm (hunk 3 above)
Comment (afishbeck) [reply esp/scm/ws_resources.ecm:43]: Maybe call it something like ExternalIPStatus?  You could make it an enum.  We also don't need to follow the kubectl formatting exactly... for example we could remove the '<','>'.

Comment (wangkx) [reply esp/scm/ws_resources.ecm:43]: Will do

**Review by afishbeck — APPROVED** [review]

File: esp/scm/ws_resources.ecm (hunk 1 above)
Comment (GordonSmith) [reply esp/scm/ws_resources.ecm:41]: @wangkx from the clients point of view, there isn't any need to report on any IPs or Ports that are "inside the cluster" - however ops may want it in the topology info?

**Comment by wangkx** [comment]
@GordonSmith I just removed the NodePort and ClusterIPs based on our discussion. Please review.

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**